client/state/query-client.ts

```diff
@@ -9,20 +13,37 @@ import {
 	storePersistedStateItem,
 } from 'calypso/state/persisted-state';
 import { shouldDehydrateQuery } from './should-dehydrate-query';
+import type { DebouncedFunc } from '@wordpress/compose';
+
+type ThrotthledPersister = {
+	persistClient: DebouncedFunc< ( state: PersistedClient ) => Promise< unknown > >;
+} & Omit< Persister, 'persistClient' >;
 
 export async function createQueryClient( userId?: number ): Promise< QueryClient > {
 	await loadPersistedState();
 	const queryClient = new QueryClient( {
 		defaultOptions: { queries: { gcTime: MAX_AGE } },
 	} );
-	await hydrateBrowserState( queryClient, userId );
+	const persister = await hydrateBrowserState( queryClient, userId );
+
+	if ( typeof window !== 'undefined' ) {
+		// Persist the cache to local storage before the browser is unloaded
+		window.addEventListener( 'beforeunload', () => {
+			if ( persister ) {
+				// flush the debouncer so the next persist is immediate
+				persister.persistClient.flush();
+				persistQueryClient( { queryClient, persister: persister as unknown as Persister } );
+			}
+		} );
```

Comment:
This will misbehave when the user is logging out. Then we clear the local storage and need to prevent the in-memory state to be saved on `beforeunload` right after it's been cleared.

Look at #53986 where this was fixed for the Redux state. In the `boot` function, after creating the query client with `createQueryClient`, you also need to register a "disable React Query persistence" callback with `onDisablePersistence`.

Then the `MeSidebar.onSignOut` handler will call `disablePersistence`, which will execute these registered callbacks, and only then will it call `await clearStore()`.

Reply:
Thanks for the suggestion @jsnajdr 

I implemented this:

```
	onDisablePersistence( () => {
		queryClient.removeQueries();
		queryClient.clear();
		queryClient.setDefaultOptions( { queries: { cacheTime: 0 } } );
	} );
```

From the documentation of react-query it wasn't 100% clear to me what the difference is between calling `removeQueries` with no arguments and `clear`. But better safe than sorry? It seems the cache is cleared after the first.

![CleanShot 2024-03-20 at 11 17 53@2x](https://github.com/Automattic/wp-calypso/assets/528287/60c29397-1e9a-41b1-bd1b-2b2ab30c54a2)

I checked IndexedDB after logging out, and there are no queries persisted, but the key is still there. Should I implement removing the key as well inside onDisablePersistence?

![CleanShot 2024-03-20 at 11 20 49@2x](https://github.com/Automattic/wp-calypso/assets/528287/31cafdb1-d55b-4d29-98c2-5fd638233830)

Reply:
> it wasn't 100% clear to me what the difference is between calling removeQueries with no arguments and clear.

Looking at the sources, I see that `removeQueries` removes matching queries from the cache. If all queries match, they will be all removed from the cache in a `for` loop. `clear` just clears the cache with one `queryCache.clear()` call. And it also clears the `mutationCache`.

This means that you only need to call `queryClient.clear()`, `queryClient.removeQueries()` is redundant.

> I checked IndexedDB after logging out, and there are no queries persisted, but the key is still there.

This is not bad, the only information we store is "user 99412650 was logged in here", but completely stopping the persister would be better. `persistQueryClient` should return an unsubscribe function. Currently we're calling `await persistQueryClient()` but that's probably a bug -- the API has changed between recent major versions, and we didn't update the usage. I think it now returns an array of `[ unsubscribeFn, promise ]`.

Then we probably need to change `createQueryClient` to return both the client and the unsubscribe function, too.

If "doing it right" prevents you from shipping a feature, then the PR is good enough as it is, but it would be nice to circle back sooner or later.

Reply:
No, I want to do it right. So, I'll be making those changes 💪

Reply:
Perfect, let me know when you have a new version ready for review 🙂